### PR TITLE
fix(#8589): don't check form expression on edit for 4.3

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

medic/cht-core#8589

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8589-430-backport/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8589-430-backport/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8589-430-backport/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
I dislike all of this. :(

File: webapp/src/ts/services/enketo.service.ts [hunk 1]
```diff
@@ -901,16 +886,38 @@ interface XmlFormContext {
...
+  contactSummary? :Record<string, any>;
+
+  constructor(selector?, type?, formDoc?, data?) {
+    this.selector = selector;
+    this.type = type;
+    this.formDoc = formDoc;
+    this.instanceData = data;
+  }
+
+  evaluateExpression() {
```
Comment (garethbowen) [inline webapp/src/ts/services/enketo.service.ts:912]: I think this name should be changed to `shouldEvaluateExpression` or similar.... it reads like it's going to actually do the evaluation.

File: webapp/src/ts/services/enketo.service.ts [hunk 2]
```diff
@@ -434,11 +434,15 @@ export class EnketoService {
...
     return this.xmlFormsService.canAccessForm(
-      formDoc,
-      user,
-      { doc: instanceData?.contact, contactSummary: contactSummary?.context },
+      formContext.formDoc,
+      formContext.userContact,
+      {
+        doc: typeof formContext.instanceData !== 'string' && formContext.instanceData?.contact,
+        contactSummary: formContext.contactSummary?.context,
+        evaluateExpression: formContext.evaluateExpression(),
```
Comment (garethbowen) [inline webapp/src/ts/services/enketo.service.ts:444]: Same naming concern here... should be a noun not a verb.

File: webapp/src/ts/services/enketo.service.ts [hunk 3]
```diff
@@ -901,16 +886,38 @@ interface XmlFormContext {
...
+
+  constructor(selector?, type?, formDoc?, data?) {
+    this.selector = selector;
+    this.type = type;
+    this.formDoc = formDoc;
+    this.instanceData = data;
+  }
+
+  evaluateExpression() {
+    if (this.type === 'report' && this.editing) {
```
Comment (garethbowen) [inline webapp/src/ts/services/enketo.service.ts:913]: Wouldn't this also apply when editing contacts?

**Review by garethbowen — APPROVED** [review]
Some nits, but otherwise you're good to go.

I think we can improve on this on `master` but for a minimal change this is good. Thanks for doing it!

File: webapp/src/ts/services/enketo.service.ts (hunk 3 above)
Comment (dianabarsan) [reply webapp/src/ts/services/enketo.service.ts:913]: I'm worried that that might be desired behavior?

Comment (dianabarsan) [reply webapp/src/ts/services/enketo.service.ts:913]: Coming back to this: the contact add and edit forms are different, so the expression can be different between them. I'm tempted to think that we should allow blocking contact edits depending on expression.

**Pull request merged**